sys/linux/dev_kvm.txt

```diff
@@ -196,6 +196,9 @@ syz_kvm_setup_cpu$x86(fd fd_kvmvm, cpufd fd_kvmcpu, usermem vma[24], text ptr[in
 syz_kvm_setup_cpu$arm64(fd fd_kvmvm, cpufd fd_kvmcpu, usermem vma[24], text ptr[in, array[kvm_text_arm64, 1]], ntext len[text], flags const[0], opts ptr[in, array[kvm_setup_opt_arm64, 1]], nopt len[opts])
 syz_kvm_setup_cpu$ppc64(fd fd_kvmvm, cpufd fd_kvmcpu, usermem vma[24], text ptr[in, array[kvm_text_ppc64, 1]], ntext len[text], flags flags[kvm_setup_flags_ppc64], opts ptr[in, array[kvm_setup_opt_ppc64, 1]], nopt len[opts])
 
+kvm_num_irqs = 32, 64, 128, 256, 512
+syz_kvm_vgic_v3_setup$arm64(fd fd_kvmvm, ncpus const[1], nirqs flags[kvm_num_irqs])
```

Comment:
Don't we want to test other ncpus? Why to we need to pass 1? The syscall could just use 1.

Reply:
The problem right now is that syz_kvm_setup_cpu() doesn't know about the number of CPUs in the system. So in order to successfully test e.g. ncpus=2 we'll need to:
* call KVM_CREATE_VCPU at least twice
* call syz_kvm_vgic_v3_setup() with ncpus=2 (not sure what happens if ncpus doesn't exactly match the number of CPUs in the VM)
* call syz_kvm_setup_cpu() for two VCPUs, passing two different CPU IDs to them

I am afraid that most of the time we'll be incorrectly guessing the number of CPUs.
Let me check though, maybe we can get the number of VCPUs for a given VM somehow.

Reply:
You mean that ncpus=1 is more likely to match the actual number of CPUs?
But then if it will fail with incorrect ncpus, perhaps syz_kvm_vgic_v3_setup should detect the current number of cpus itself. It's very easy to create more than 1 cpu for the fuzzer (it does not need to call syz_kvm_vgic_v3_setup), just more than 1 create VM ioctls.

Reply:
What I didn't know is that the second parameter of KVM_CREATE_VCPU is CPU ID, and duplicate IDs aren't possible (the ioctl returns EEXIST).
So we cannot create more than 3 VCPUs in a single VM (because the description limits CPU IDs to 0:2), and therefore it is fine to always pass ncpus=4 here. What do you think?

Reply:
This really depends on what happens if ncpus does not match actual value.
Does it have to match precisely? Or passing larger is permitted? Is passing smaller value permitted?

Reply:
I think passing a larger value is permitted. As for passing a smaller value, the VM might fail to allocate enough space for the distributor/redistributor in the VCPU, which is good.
That being said, we can probably go ahead and make ncpus take values within the range of 0:4.

Reply:
Sounds reasonable.